Zend/zend_vm_def.h

```diff
@@ -1929,6 +1929,30 @@ ZEND_VM_HANDLER(84, ZEND_FETCH_DIM_W, VAR|CV, CONST|TMPVAR|UNUSED|NEXT|CV)
 	USE_OPLINE
 	zval *container;
 
+	SAVE_OPLINE();
+	container = GET_OP1_ZVAL_PTR_PTR_UNDEF(BP_VAR_W);
+	zend_fetch_dimension_address_W(container, GET_OP2_ZVAL_PTR_UNDEF(BP_VAR_R), OP2_TYPE OPLINE_CC EXECUTE_DATA_CC);
+	FREE_OP2();
+	if (opline->extended_value == ZEND_FETCH_DIM_REF) {
+ZEND_VM_C_LABEL(gc_mark_collectable_again):
+		if (EXPECTED(Z_TYPE_P(container) == IS_ARRAY)) {
+			gc_mark_collectable(Z_COUNTED_P(container));
+		} else if (Z_TYPE_P(container) == IS_REFERENCE) {
+			container = Z_REFVAL_P(container);
+			ZEND_VM_C_GOTO(gc_mark_collectable_again);
+		}
+	}
+	if (OP1_TYPE == IS_VAR) {
+		FREE_VAR_PTR_AND_EXTRACT_RESULT_IF_NECESSARY(opline->op1.var);
+	}
+	ZEND_VM_NEXT_OPCODE_CHECK_EXCEPTION();
```

Comment:
zend_fetch_dimension_address_W() is implemented through zend_hash_lookup() that already takes care about calling gc_mark_collectable(). This chunk is probably necessary only for fast path of ZEND_HASH_INDEX_LOOKUP().